### DISCUSSION TITLE
2 way audio?

### DISCUSSION THREAD

**Discussion by KendallStauffer** [tsightler/ring-mqtt#312] [General Discussion]
I have read everything and dont see any mention of 2 way audio like the ring app has, is that possible?

Also the picture Glance card works great but would be perfect if there was a way to have the live image (when you click the card) be a webrtc image, but keep everything else the way it is (seeing what caused the motion Is my main use)

**Comment by tsightler** [comment]
There's no two way audio support in HA that I know of.  Certainly there's none in this addon and will likely never be.  The primary goal of this project is not to be a Ring app replacement.  The goal is automation, not interaction.

Regarding glance card, what's wrong with the native HA card?  If you enable low-latency HLS it's probably only 2 seconds extra delay vs Webrtc  and webrtc should eventually be native in HA (the framework was added in one of the recent versions, but there's more work to-do.  Regardless, these are outside of this project so probably best asked on HA forums vs here.

**Reply by KendallStauffer** [reply]
I did enable HLS and it works on my non ring cameras and I agree is fast enough, but I am not having That result with the ring camera, do you know anyway I can troubleshoot what I am doing wrong? Mine is more likely 9 or 10 seconds behind.  Everything else working fine, live stream switch stays off except when being accessed

**Reply by tsightler** [reply]
This is really an HA question for the most part so probably best to ask on the HA forums as I'm far from an expert on HA (I really barely use it, only maintain the addon because otherwise people made their own and still asked me questions about it).  You can check the delay added via the addon by connecting to the stream with an RTSP client like VLC, but normally that delay should be no more than 1 second or so.  I use the following settings on my HA instance for Ring camera:

> stream:
>   ll_hls: true
>   part_duration: 0.75
>   segment_duration: 2.25

This gives me about 3 seconds delay using native LL-HLS in the HA UI.  Not perfect but more reliable than the WebRTC card with 7 cameras and seems good enough to me.  Ring cameras will never have as little delay as local cameras since all streams have to go to Ring cloud servers and back.

**Comment by tsightler** [comment]
Also, this is not a feature request so I'm moving this to the General discussion forum.

**Comment by KendallStauffer** [comment]
I wasn’t thinking so much about emulating ring app as having an alarm panel
and what features those have.

Kendall Stauffer


On Jan 26, 2022, at 5:53 PM, tsightler ***@***.***> wrote:

﻿

Also, this is not a feature request so I'm moving this to the General
discussion forum.

—
Reply to this email directly, view it on GitHub
<https://github.com/tsightler/ring-mqtt/discussions/312#discussioncomment-2054721>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/ABJ2HBBQOWFSWNHZWNWXQVLUYB3NXANCNFSM5M4IR6WA>
.
Triage notifications on the go with GitHub Mobile for iOS
<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
or Android
<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

You are receiving this because you authored the thread.Message ID:
***@***.***>

**Comment by tsightler** [comment]
There's no two-way audio support in HA that I know of and there's also none in this code.  While the underlying API does support this, it adds a lot of complication to deal with and I have no plans to add this any time in the foreseeable future.  If native support for two-way audio makes it to HA at some point I might reconsider.

**Comment by KendallStauffer** [comment]
Thanks much!!

**Comment by KendallStauffer** [comment]
Thanks so much for your quick response. I will check with VLC later today.

I did have my stream setting a little different, I had 6 instead of 2.25
but it didnt make any different on the card. Do you set the stream for each
object, or just once (I have it just once global in my configuration )

   I have a Raspberry pi 4b with 8g RAM and 64g of the faster io memory
card (A2)
Should that be fast enough?

**Reply by tsightler** [reply]
I'm afraid I don't have any more ideas.  IMO there's no way you should be seeing >5 seconds of latency.  Even with standard HLS it should only be about 6-7 seconds.  

The hardware feels like way more than would ever be needed unless it's doing a ton of other stuff.  I run as many as 7 ring cameras concurrently on my RPi4b with only 4GB of RAM and overall utilization remains reasonable.

**Comment by KendallStauffer** [comment]
Thanks you are helping by letting me know what to expect!!

**Reply by tsightler** [reply]
Well, good luck.  Once you try using VLC you'll get some idea of the baseline latency that is added just going through the addon and RTSP server (the actual stream from Ring servers is a SIP session which is piped out by ring-client-api to an ffmpeg session that forwards the stream to the RTSP server for easier consumption by RTSP clients).  Although I've tried to tune this for minimum latency, it's still a lot of hops, including the RTP stream being processed by NodeJS, which isn't the most efficient for that type of work.  When testing with VLC I've seen about 1 second of latency if I override the default 1 second buffer that VLC adds to 0 seconds.

HLS typically uses segments split in the keyframes of the stream and has to buffer 3 segments so the latency is dependent on how often the stream includes a keyframe.  In the streams I get from Ring for my cameras they generally have a keyframe interval of 2 seconds so 3 buffered frames is 6 seconds of video.  This is why standard HLS has 6-7 seconds of latency.

Low latency HLS breaks each segment down into small parts and should cut the latency to basically the equivalent of 1 segment, at the expense of much higher processing overhead (generally only works well over HTTP/2 since that protocol uses a single persistent connection instead of many TCP connections).

It's completely possible that not all cameras or servers in Ring use the exact same settings so you may be seeing larger segments due to less keyframe frequency, although the fact that you see no difference between standard HLS and LL-HLS seems to indicate that something is just not working correctly.  In all cases that I can think of there should be a dramatic difference between these two.

As to WebRTC, I'm hopeful that HA might eventually have support in the native Ring component eventually.  Ring is supporting WebRTC on the server side even today (their web app uses it) so I can envision a day where the native Ring integration can just use the API to present the WebRTC stream directly within HA, without the need for this addon.

In the interim, it might be possible to use something like a conditional card to do mostly what you want, but I haven't tried this.  I could think that you could add a boolean value which could be added to the picture glance card and, when you click it, it exposes a conditional card that has the WebRTC image.  There may even be better ways but these are better left to people who are HA experts, which I certainly am not.
